### PR TITLE
HPU & TPU doesn't support torch.inference_mode

### PR DESCRIPTION
**Pull request by jerome-habana**
Signed-off-by: Jerome <janand@habana.ai>

## What does this PR do?

https://github.com/PyTorchLightning/pytorch-lightning/pull/12715/ introduced torch.inference_mode without checking all backends. This resulted in the following error on habana backends
"RuntimeError: Cannot set version_counter for inference tensor"

This PR disables inference_mode and uses no_grad mode for eval/predict stages with hpu backend. Once hpu impl adds support for inference_mode, this check will be removed.

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2802,10 +2802,12 @@ def configure_optimizers(self):
...
 @contextmanager
 def _evaluation_context() -> Generator:
-    # inference mode is not supported with gloo backend (#9431)
+    # inference mode is not supported with gloo backend (#9431) and hpu backend
     context_manager_class = (
         torch.inference_mode
-        if _TORCH_GREATER_EQUAL_1_9 and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        if _TORCH_GREATER_EQUAL_1_9
+        and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        and not (_HPU_AVAILABLE)
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:2810]:
```suggestion
        and not isinstance(self.accelerator, HPUAccelerator)
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2802,10 +2802,12 @@ def configure_optimizers(self):
...
 @contextmanager
 def _evaluation_context() -> Generator:
-    # inference mode is not supported with gloo backend (#9431)
+    # inference mode is not supported with gloo backend (#9431) and hpu backend
     context_manager_class = (
         torch.inference_mode
-        if _TORCH_GREATER_EQUAL_1_9 and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        if _TORCH_GREATER_EQUAL_1_9
+        and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        and not isinstance(self.accelerator, HPUAccelerator)
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:2810]:
```suggestion
        and not _HPU_AVAILABLE
```

**Review by kaushikb11 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by kaushikb11** [comment]
@awaelchli @tchaton

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2802,10 +2802,12 @@ def configure_optimizers(self):
...
 @contextmanager
 def _evaluation_context() -> Generator:
-    # inference mode is not supported with gloo backend (#9431)
+    # inference mode is not supported with gloo backend (#9431) and hpu backend
     context_manager_class = (
         torch.inference_mode
-        if _TORCH_GREATER_EQUAL_1_9 and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        if _TORCH_GREATER_EQUAL_1_9
+        and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        and not _HPU_AVAILABLE
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2810]: I'd rather suggest conditioning it on if HPU is used or not for the current run?

File: docs/source/accelerators/hpu_basic.rst [hunk 4]
```diff
@@ -81,3 +81,4 @@ Known limitations
 * Multiple optimizers are not supported.
 * `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__ is not supported.
 * :class:`~pytorch_lightning.callbacks.device_stats_monitor.DeviceStatsMonitor` is not supported.
+* torch.inference_mode() is not supported
```
Comment (rohitgr7) [inline docs/source/accelerators/hpu_basic.rst:84]:
```suggestion
* :func:`torch.inference_mode` is not supported
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -2801,11 +2801,13 @@ def configure_optimizers(self):
 
 
 @contextmanager
-def _evaluation_context() -> Generator:
-    # inference mode is not supported with gloo backend (#9431)
+def _evaluation_context(accelerator) -> Generator:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2804]:
```suggestion
def _evaluation_context(accelerator: Accelerator) -> Generator:
```

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -2801,11 +2801,13 @@ def configure_optimizers(self):
...
+    # inference mode is not supported with gloo backend (#9431) and hpu backend
     context_manager_class = (
         torch.inference_mode
-        if _TORCH_GREATER_EQUAL_1_9 and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        if _TORCH_GREATER_EQUAL_1_9
+        and not (dist.is_initialized() and dist.get_backend() == "gloo")
+        and not (isinstance(accelerator, HPUAccelerator))
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2810]:
```suggestion
        and not isinstance(accelerator, HPUAccelerator)
```

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — COMMENTED** [review]
go

**Pull request merged**